File: src/GameOfLife/Renderer.cs
using System.Text;

namespace GameOfLife;

/// <summary>
/// Console renderer that animates the Game of Life grid in the terminal.
/// </summary>
public class Renderer
{
    private const char AliveChar = '\u2588'; // Full block
    private const char DeadChar = ' ';

    public int ViewportWidth { get; set; } = 80;
    public int ViewportHeight { get; set; } = 24;
    public int OffsetX { get; set; }
    public int OffsetY { get; set; }

    /// <summary>
    /// Renders the grid to a string for the current viewport.
    /// </summary>
    public string Render(Grid grid)
    {
        var sb = new StringBuilder();

        for (int y = OffsetY; y < OffsetY + ViewportHeight; y++)
        {
            for (int x = OffsetX; x < OffsetX + ViewportWidth; x++)
            {
                sb.Append(grid.IsAlive(new Cell(x, y)) ? AliveChar : DeadChar);
            }
            if (y < OffsetY + ViewportHeight - 1)
                sb.AppendLine();
        }

        return sb.ToString();
    }

    /// <summary>
    /// Renders the grid to the console, clearing the screen first.
    /// </summary>
    public void RenderToConsole(Grid grid)
    {
        Console.SetCursorPosition(0, 0);
        Console.Write(Render(grid));
        Console.WriteLine();
        Console.WriteLine($"Generation: {grid.Generation}  Population: {grid.Population}    ");
    }

    /// <summary>
    /// Animates the grid in the console for the given number of generations.
    /// </summary>
    public void Animate(Grid grid, int generations = int.MaxValue, int delayMs = 100, CancellationToken ct = default)
    {
        Console.Clear();
        Console.CursorVisible = false;

        // Auto-center viewport on the pattern
        CenterOnPattern(grid);

        try
        {
            for (int i = 0; i < generations && !ct.IsCancellationRequested; i++)
            {
                RenderToConsole(grid);
                Thread.Sleep(delayMs);
                grid.Step();
            }
        }
        finally
        {
            Console.CursorVisible = true;
        }
    }

    /// <summary>
    /// Centers the viewport on the current live cells.
    /// </summary>
    public void CenterOnPattern(Grid grid)
    {
        var bounds = grid.GetBounds();
        if (bounds == null) return;

        var (minX, minY, maxX, maxY) = bounds.Value;
        int centerX = (minX + maxX) / 2;
        int centerY = (minY + maxY) / 2;

        OffsetX = centerX - ViewportWidth / 2;
        OffsetY = centerY - ViewportHeight / 2;
    }
}

File: src/game_of_life/renderer.clj
(ns game-of-life.renderer
  "Console renderer that prints the Game of Life grid to the terminal."
  (:require [game-of-life.grid :as grid]))

(def ^:private alive-char \u2588)  ; Full block
(def ^:private dead-char \space)

(defn render
  "Renders the grid to a string for the given viewport.
   Returns a string representation of the grid."
  [live-cells {:keys [width height offset-x offset-y]
               :or {width 80 height 24 offset-x 0 offset-y 0}}]
  (let [sb (StringBuilder.)]
    (doseq [y (range offset-y (+ offset-y height))]
      (doseq [x (range offset-x (+ offset-x width))]
        (.append sb (if (contains? live-cells [x y]) alive-char dead-char)))
      (when (< y (+ offset-y height -1))
        (.append sb \newline)))
    (.toString sb)))

(defn center-viewport
  "Returns viewport options centered on the current live cells."
  [live-cells width height]
  (if-let [[min-x min-y max-x max-y] (grid/bounds live-cells)]
    (let [center-x (quot (+ min-x max-x) 2)
          center-y (quot (+ min-y max-y) 2)]
      {:width width
       :height height
       :offset-x (- center-x (quot width 2))
       :offset-y (- center-y (quot height 2))})
    {:width width :height height :offset-x 0 :offset-y 0}))

(defn render-to-console
  "Renders the grid to the console, clearing the screen first."
  [live-cells generation viewport]
  (print "\033[H")  ; Move cursor to top-left
  (println (render live-cells viewport))
  (printf "Generation: %d  Population: %d    %n" generation (count live-cells))
  (flush))

(defn animate
  "Animates the grid in the console for the given number of generations."
  [initial-cells & {:keys [generations delay-ms width height]
                    :or {generations Integer/MAX_VALUE
                         delay-ms 100
                         width 80
                         height 24}}]
  (print "\033[2J")  ; Clear screen
  (let [viewport (center-viewport initial-cells width height)]
    (loop [cells initial-cells
           gen 0]
      (when (< gen generations)
        (render-to-console cells gen viewport)
        (Thread/sleep delay-ms)
        (recur (grid/step cells) (inc gen))))))
